File: src/com/castlefrog/agl/domains/havannah/HavannahSimulator.java
package com.castlefrog.agl.domains.havannah;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

import com.castlefrog.agl.AbstractSimulator;
import com.castlefrog.agl.IllegalActionException;
import com.castlefrog.agl.TurnType;

public final class HavannahSimulator extends AbstractSimulator<HavannahState, HavannahAction> {
    private static final int N_AGENTS = 2;
    private static final int[] REWARDS_BLACK_WINS = new int[] { 1, -1 };
    private static final int[] REWARDS_WHITE_WINS = new int[] { -1, 1 };
    private static final int[] REWARDS_NEUTRAL = new int[] { 0, 0 };

    private TurnType turnType_;
    /** length of a side of board */
    private final int base_;
    /** longest row of hexagons on board (always odd) */
    private final int size_;
    /** number of locations on board */
    private final int nLocations_;
    private final int[][] corners_;
    private final int[][][] sides_;

    private HavannahSimulator(HavannahState state) {
        legalActions_ = new ArrayList<>();
        legalActions_.add(new ArrayList<HavannahAction>());
        legalActions_.add(new ArrayList<HavannahAction>());
        setState(state);
        base_ = state.getBase();
        size_ = state.getSize();
        nLocations_ = state.getNLocations();
        corners_ = state.getCorners();
        sides_ = state.getSides();
    }

    private HavannahSimulator(HavannahSimulator simulator,
                              int[][] corners,
                              int[][][] sides) {
        super(simulator);
        base_ = simulator.getState().getBase();
        size_ = simulator.getState().getSize();
        nLocations_ = simulator.getState().getNLocations();
        corners_ = corners;
        sides_ = sides;
        turnType_ = getTurnType();
    }

    public HavannahSimulator copy() {
        return new HavannahSimulator(this, corners_, sides_);
    }

    public static HavannahSimulator create(int boardSize, TurnType turnType) {
        return new HavannahSimulator(getInitialState(boardSize));
    }

    public void setState(HavannahState state) {
        state_ = state;
        computeRewards();
        computeLegalActions();
    }

    public void stateTransition(List<HavannahAction> actions) {
        HavannahAction action = actions.get(state_.getAgentTurn());
        if (!legalActions_.get(state_.getAgentTurn()).contains(action)) {
            throw new IllegalActionException(action, state_);
        }
        state_.setLocation(action.getX(), action.getY(), state_.getAgentTurn() + 1);
        state_.setAgentTurn((state_.getAgentTurn() + 1) % N_AGENTS);
        computeRewards(action);
        computeLegalActions(action);
    }

    private void computeLegalActions() {
        computeLegalActions(null);
    }

    private void computeLegalActions(HavannahAction prevAction) {
        if (rewards_ == REWARDS_NEUTRAL) {
            int agentTurn = state_.getAgentTurn();
            int otherTurn = (agentTurn + 1) % N_AGENTS;
            legalActions_.set(agentTurn, legalActions_.get(otherTurn));
            legalActions_.set(otherTurn, new ArrayList<HavannahAction>());
            List<HavannahAction> legalActions = legalActions_.get(agentTurn);
            if (prevAction != null && nLocations_ - legalActions.size() >= 2) {
                legalActions.remove(prevAction);
            } else {
                legalActions.clear();
                int count = 0;
                HavannahAction tempAction = null;
                for (int y = 0; y < size_; y += 1) {
                    int xMin = 0;
                    int xMax = size_;
                    if (y >= base_) {
                        xMin = y - base_ + 1;
                    } else {
                        xMax = base_ + y;
                    }

                    for (int x = xMin; x < xMax; x += 1) {
                        if (state_.getLocation(x, y) == 0) {
                            legalActions.add(HavannahAction.valueOf(x, y));
                        } else if (state_.getAgentTurn() == 1 && count == 0) {
                            count = 1;
                            tempAction = HavannahAction.valueOf(x, y);
                        } else if (state_.getAgentTurn() == 1 && count == 1) {
                            count = 2;
                            tempAction = null;
                        }
                    }
                }
                if (tempAction != null) {
                    legalActions.add(tempAction);
                }
            }
        } else {
            for (List<HavannahAction> legalActions: legalActions_) {
                legalActions.clear();
            }
        }
    }

    private void computeRewards() {
        computeRewards(null);
    }

    private void computeRewards(HavannahAction prevAction) {
        byte[][] locations = state_.getLocations();
        boolean[][] visited = new boolean[size_][size_];
        int yMin = 0;
        int xMin = 0;
        int yMax = size_;
        int xMax = size_;
        if (prevAction != null) {
            xMin = prevAction.getX();
            yMin = prevAction.getY();
            xMax = xMin + 1;
            yMax = yMin + 1;
        }
        for (int y = yMin; y < yMax; y += 1) {
            for (int x = xMin; x < xMax; x += 1) {
                // Checks: non empty location - hasn't been visited
                if (locations[x][y] != 0 && !visited[x][y]) {
                    int result = dfsCornersSides(x, y, locations, visited);
                    // count corners
                    int corners = 0;
                    for (int k = 0; k < 6; k += 1) {
                        if (result % 2 == 1) {
                            corners += 1;
                        }
                        result >>= 1;
                    }
                    // count sides
                    int sides = 0;
                    for (int k = 0; k < 6; k += 1) {
                        if (result % 2 == 1) {
                            sides += 1;
                        }
                        result >>= 1;
                    }
                    if (corners >= 2 || sides >= 3) {
                        if (locations[x][y] == 1) {
                            rewards_ = REWARDS_BLACK_WINS;
                            return;
                        } else {
                            rewards_ = REWARDS_WHITE_WINS;
                            return;
                        }
                    }
                }
            }
        }

        locations = state_.getLocations();
        visited = new boolean[size_][size_];
        for (int y = 0; y < locations.length; y += 1) {
            xMin = 0;
            xMax = size_;
            if (y >= base_) {
                xMin = y - base_ + 1;
            } else {
                xMax = base_ + y;
            }
            for (int x = xMin; x < xMax; x += 1) {
                if (locations[x][y] == 0 || locations[x][y] == state_.getAgentTurn() + 1) {
                    locations[x][y] = 1;
                } else {
                    locations[x][y] = 0;
                }
            }
        }

        yMin = 0;
        xMin = 0;
        yMax = size_;
        xMax = size_;
        if (prevAction != null) {
            xMin = Math.max(prevAction.getX() - 1, 0);
            yMin = Math.max(prevAction.getY() - 1, 0);
            xMax = Math.min(prevAction.getX() + 2, size_);
            yMax = Math.min(prevAction.getY() + 2, size_);
        }

        for (int y = yMin; y < yMax; y += 1) {
            for (int x = xMin; x < xMax; x += 1) {
                if (locations[x][y] != 0 && !visited[x][y]) {
                    if (dfsCornersSides(x, y, locations, visited) == 0) {
                        if (state_.getAgentTurn() == 0) {
                            rewards_ = REWARDS_WHITE_WINS;
                            return;
                        } else {
                            rewards_ = REWARDS_BLACK_WINS;
                            return;
                        }
                    }
                }
            }
        }
        rewards_ = REWARDS_NEUTRAL;
    }

    private int dfsCornersSides(int x0,
                                int y0,
                                byte[][] locations,
                                boolean[][] visited) {
        int value = 0;
        Stack<HavannahAction> stack = new Stack<>();
        stack.push(HavannahAction.valueOf(x0, y0));
        visited[x0][y0] = true;
        while (!stack.empty()) {
            HavannahAction v = stack.pop();
            int x = v.getX();
            int y = v.getY();
            value |= getCornerMask(x, y) | getSideMask(x, y);
            for (int i = -1; i <= 1; i += 1) {
                for (int j = -1; j <= 1; j += 1) {
                    int xi = x + i;
                    int yi = y + j;
                    if (i + j != 0 && xi >= 0 && yi >= 0 &&
                            xi < size_ && yi < size_ &&
                            (yi < base_ && xi < base_ + yi ||
                                yi >= base_ && xi > yi - base_)) {
                        if (!visited[xi][yi] &&
                                locations[xi][yi] == locations[x][y]) {
                            stack.push(HavannahAction.valueOf(xi, yi));
                            visited[xi][yi] = true;
                        }
                    }
                }
            }
        }
        return value;
    }

    public List<HavannahAction> getWinningConnection() {
        List<HavannahAction> connection = new ArrayList<>();
        if (rewards_ != REWARDS_NEUTRAL) {
            HavannahSimulator simulator = HavannahSimulator.create(base_, turnType_);
            HavannahState state = state_.copy();
            for (int i = 0; i < size_; i += 1) {
                for (int j = 0; j < size_; j += 1) {
                    int location = state.getLocation(i, j);
                    if (!state.isLocationEmpty(i, j) &&
                            location != state.getAgentTurn() + 1) {
                        state.setLocation(i, j, HavannahState.LOCATION_EMPTY);
                        simulator.setState(state);
                        if (!simulator.isTerminalState()) {
                            connection.add(HavannahAction.valueOf(i, j));
                            state.setLocation(i, j, location);
                        }
                    }
                }
            }
        }
        return connection;
    }

    private int getCornerMask(int x, int y) {
        for (int i = 0; i < corners_.length; i += 1) {
            if (corners_[i][0] == x && corners_[i][1] == y) {
                return 1 << i;
            }
        }
        return 0;
    }

    private int getSideMask(int x, int y) {
        for (int i = 0; i < sides_.length; i += 1) {
            for (int j = 0; j < sides_[i].length; j += 1) {
                if (sides_[i][j][0] == x && sides_[i][j][1] == y) {
                    return 1 << (i + 6);
                }
            }
        }
        return 0;
    }

    public static HavannahState getInitialState(int base) {
        return new HavannahState(base, new byte[2 * base - 1][2 * base - 1], 0);
    }

    public int getNAgents() {
        return N_AGENTS;
    }

    public TurnType getTurnType() {
        return turnType_;
    }
}

File: src/main/kotlin/com/castlefrog/agl/domains/havannah/HavannahSimulator.kt
package com.castlefrog.agl.domains.havannah

import arrow.core.Option
import com.castlefrog.agl.Simulator
import com.castlefrog.agl.domains.AdversarialRewards
import com.castlefrog.agl.domains.nextPlayerTurnSequential
import com.castlefrog.agl.util.LruCache
import kotlin.math.max
import kotlin.math.min

class HavannahSimulator(
    val base: Int = 10,
    val pieRule: Boolean = true
) : Simulator<HavannahState, HavannahAction> {

    /** keep track of prev action for performance reasons */
    private val prevActionCache = LruCache<HavannahState, HavannahAction>(1)
    private val size = 2 * base - 1
    private val actions: Array<Array<HavannahAction>> = HavannahAction.generateActions(size)

    private data class Bounds(val xMin: Int, val xMax: Int, val yMin: Int, val yMax: Int)

    private val corners: Array<IntArray>
        get() = arrayOf(
            intArrayOf(0, 0), intArrayOf(0, base - 1), intArrayOf(base - 1, 0),
            intArrayOf(base - 1, size - 1), intArrayOf(size - 1, base - 1), intArrayOf(size - 1, size - 1)
        )

    private val sides: Array<Array<IntArray>>
        get() {
            val sides = Array(6) { Array(base - 2) { IntArray(2) } }
            for (i in 0..base - 3) {
                sides[0][i][0] = 0
                sides[0][i][1] = i + 1
                sides[1][i][0] = i + 1
                sides[1][i][1] = 0
                sides[2][i][0] = i + 1
                sides[2][i][1] = base + i
                sides[3][i][0] = base + i
                sides[3][i][1] = size - 1
                sides[4][i][0] = size - 1
                sides[4][i][1] = base + i
                sides[5][i][0] = base + i
                sides[5][i][1] = i + 1
            }
            return sides
        }

    init {
        if (base < MIN_BASE) {
            throw IllegalArgumentException("Invalid board size: $base")
        }
    }

    override val initialState: HavannahState
        get() = HavannahState(base, Array(2 * base - 1) { ByteArray(2 * base - 1) }, HavannahState.TURN_BLACK)

    override fun calculateRewards(state: HavannahState): IntArray {
        val prevAction = prevActionCache[state]
        val visited = Array(size) { BooleanArray(size) }
        val bounds = if (prevAction != null) {
            val xMin = prevAction.x.toInt()
            val yMin = prevAction.y.toInt()
            val xMax = xMin + 1
            val yMax = yMin + 1
            Bounds(xMin = xMin, xMax = xMax, yMin = yMin, yMax = yMax)
        } else {
            Bounds(xMin = 0, xMax = size, yMin = 0, yMax = size)
        }
        for (y in bounds.yMin until bounds.yMax) {
            for (x in bounds.xMin until bounds.xMax) {
                // Checks: non empty location - hasn't been visited
                if (!state.isLocationEmpty(x, y) && !visited[x][y]) {
                    var result = dfsCornersSides(x, y, state, visited, corners, sides, actions)
                    // count corners
                    var nCorners = 0
                    for (k in 0..5) {
                        if (result % 2 == 1) {
                            nCorners += 1
                        }
                        result = result shr 1
                    }
                    // count sides
                    var nSides = 0
                    for (k in 0..5) {
                        if (result % 2 == 1) {
                            nSides += 1
                        }
                        result = result shr 1
                    }
                    if (nCorners >= 2 || nSides >= 3) {
                        return if (state.locations[x][y] == HavannahState.LOCATION_BLACK) {
                            AdversarialRewards.BLACK_WINS
                        } else {
                            AdversarialRewards.WHITE_WINS
                        }
                    }
                }
            }
        }

        val otherState = state.copy()
        for (y in otherState.locations.indices) {
            val xMin = if (y >= state.base) y - state.base + 1 else 0
            val xMax = if (y >= state.base) size else state.base + y
            for (x in xMin until xMax) {
                if (otherState.isLocationEmpty(
                        x,
                        y
                    ) || otherState.locations[x][y] == (otherState.agentTurn + 1).toByte()
                ) {
                    otherState.locations[x][y] = HavannahState.LOCATION_BLACK
                } else {
                    otherState.locations[x][y] = HavannahState.LOCATION_EMPTY
                }
            }
        }

        val bounds2 = if (prevAction != null) {
            val xMin = max(prevAction.x - 1, 0)
            val yMin = max(prevAction.y - 1, 0)
            val xMax = min(prevAction.x + 2, size)
            val yMax = min(prevAction.y + 2, size)
            Bounds(xMin = xMin, xMax = xMax, yMin = yMin, yMax = yMax)
        } else {
            Bounds(xMin = 0, xMax = size, yMin = 0, yMax = size)
        }

        val visited2 = Array(size) { BooleanArray(size) }
        for (y in bounds2.yMin until bounds2.yMax) {
            for (x in bounds2.xMin until bounds2.xMax) {
                if (!otherState.isLocationEmpty(x, y) && !visited2[x][y]) {
                    if (dfsCornersSides(x, y, otherState, visited2, corners, sides, actions) == 0) {
                        return if (otherState.agentTurn == HavannahState.TURN_BLACK) {
                            AdversarialRewards.WHITE_WINS
                        } else {
                            AdversarialRewards.BLACK_WINS
                        }
                    }
                }
            }
        }
        return AdversarialRewards.NEUTRAL
    }

    override fun calculateLegalActions(state: HavannahState): List<Set<HavannahAction>> {
        val legalActions = arrayListOf<MutableSet<HavannahAction>>(mutableSetOf(), mutableSetOf())
        val rewards = calculateRewards(state)
        if (rewards.contentEquals(AdversarialRewards.NEUTRAL)) {
            var count = 0
            var tempAction: HavannahAction? = null
            for (y in 0 until size) {
                val xMin = if (y >= state.base) y - state.base + 1 else 0
                val xMax = if (y >= state.base) size else state.base + y

                for (x in xMin until xMax) {
                    if (state.isLocationEmpty(x, y)) {
                        legalActions[state.agentTurn.toInt()].add(actions[x][y])
                    } else if (pieRule && count == 0 && state.agentTurn == HavannahState.TURN_WHITE) {
                        count = 1
                        tempAction = actions[x][y]
                    } else if (pieRule && count == 1 && state.agentTurn == HavannahState.TURN_WHITE) {
                        count = 2
                        tempAction = null
                    }
                }
            }
            if (tempAction != null) {
                legalActions[state.agentTurn.toInt()].add(tempAction)
            }
        }
        return legalActions
    }

    override fun stateTransition(state: HavannahState, actions: List<Option<HavannahAction>>): HavannahState {
        val action = actions[state.agentTurn.toInt()].orNull()
        val legalActions = calculateLegalActions(state)
        if (action === null || !legalActions[state.agentTurn.toInt()].contains(action)) {
            throw IllegalArgumentException("Illegal action, $action, from state, $state")
        }
        prevActionCache[state] = action
        state.locations[action.x.toInt()][action.y.toInt()] = (state.agentTurn + 1).toByte()
        state.agentTurn = nextPlayerTurnSequential(state.agentTurn.toInt(), NUMBER_OF_PLAYERS).toByte()
        return state
    }

    override fun numberOfPlayers(): Int = NUMBER_OF_PLAYERS

    companion object {
        private const val MIN_BASE = 2
        private const val NUMBER_OF_PLAYERS = 2

        private fun dfsCornersSides(
            x0: Int,
            y0: Int,
            state: HavannahState,
            visited: Array<BooleanArray>,
            corners: Array<IntArray>,
            sides: Array<Array<IntArray>>,
            actions: Array<Array<HavannahAction>>
        ): Int {
            val size = 2 * state.base - 1
            var value = 0
            val stack = ArrayDeque<HavannahAction>()
            stack.addFirst(actions[x0][y0])
            visited[x0][y0] = true
            while (!stack.isEmpty()) {
                val v = stack.removeFirst()
                val x = v.x.toInt()
                val y = v.y.toInt()
                value = value or (getCornerMask(x, y, corners) or getSideMask(x, y, sides))
                for (i in -1..1) {
                    for (j in -1..1) {
                        val xi = x + i
                        val yi = y + j
                        if (i + j != 0 && xi >= 0 && yi >= 0 &&
                            xi < size && yi < size &&
                            (yi < state.base && xi < state.base + yi || yi >= state.base && xi > yi - state.base)
                        ) {
                            if (!visited[xi][yi] && state.locations[xi][yi] == state.locations[x][y]) {
                                stack.addFirst(actions[xi][yi])
                                visited[xi][yi] = true
                            }
                        }
                    }
                }
            }
            return value
        }

        private fun getCornerMask(x: Int, y: Int, corners: Array<IntArray>): Int {
            return (corners.indices)
                .firstOrNull { corners[it][0] == x && corners[it][1] == y }
                ?.let { 1 shl it }
                ?: 0
        }

        private fun getSideMask(x: Int, y: Int, sides: Array<Array<IntArray>>): Int {
            for (i in sides.indices) {
                for (j in sides[i].indices) {
                    if (sides[i][j][0] == x && sides[i][j][1] == y) {
                        return 1 shl i + 6
                    }
                }
            }
            return 0
        }
    }
}
